framework/src/utils/RankTwoTensor.C

```diff
@@ -889,6 +889,259 @@ RankTwoTensorTempl<T>::symmetricEigenvaluesEigenvectors(std::vector<T> & eigvals
+      {
+        f = d[j];
+        V(j, i) = f;
+        g = e[j] + V(j, j) * f;
+        for (int k = j + 1; k <= i - 1; k++)
+        {
+          g += V(k, j) * d[k];
+          e[k] += V(k, j) * f;
+        }
+        e[j] = g;
+      }
+      f = 0.0;
+      for (int j = 0; j < i; j++)
+      {
+        e[j] /= h;
+        f += e[j] * d[j];
+      }
+      T hh = f / (h + h);
+      for (int j = 0; j < i; j++)
+        e[j] -= hh * d[j];
+      for (int j = 0; j < i; j++)
+      {
+        f = d[j];
+        g = e[j];
+        for (int k = j; k <= i - 1; k++)
+          V(k, j) -= (f * e[k] + g * d[k]);
+        d[j] = V(i - 1, j);
+        V(i, j) = 0.0;
+      }
+    }
+    d[i] = h;
+  }
+
+  // Accumulate transformations.
+  for (int i = 0; i < N - 1; i++)
+  {
+    V(N - 1, i) = V(i, i);
+    V(i, i) = 1.0;
+    T h = d[i + 1];
+    if (h != 0.0)
+    {
+      for (int k = 0; k <= i; k++)
+        d[k] = V(k, i + 1) / h;
+      for (int j = 0; j <= i; j++)
+      {
+        T g = 0.0;
+        for (int k = 0; k <= i; k++)
+          g += V(k, i + 1) * V(k, j);
+        for (int k = 0; k <= i; k++)
+          V(k, j) -= g * d[k];
+      }
+    }
+    for (int k = 0; k <= i; k++)
+      V(k, i + 1) = 0.0;
+  }
+  for (int j = 0; j < N; j++)
+  {
+    d[j] = V(N - 1, j);
+    V(N - 1, j) = 0.0;
+  }
+  V(N - 1, N - 1) = 1.0;
+  e[0] = 0.0;
+}
+
+template <typename T>
+void
+RankTwoTensorTempl<T>::tridiagonalQLDecomposition(RankTwoTensorTempl<T> & V,
+                                                  std::vector<T> & d,
+                                                  T e[3]) const
+{
+  int N = 3;
+  //  This is derived from the Algol procedures tql2, by
+  //  Bowdler, Martin, Reinsch, and Wilkinson, Handbook for
+  //  Auto. Comp., Vol.ii-Linear Algebra, and the corresponding
+  //  Fortran subroutine in EISPACK.
+
+  for (int i = 1; i < N; i++)
+    e[i - 1] = e[i];
+  e[N - 1] = 0.0;
+
+  T f = 0.0;
+  T tst1 = 0.0;
+  T eps = std::pow(2.0, -52.0);
+  for (int l = 0; l < N; l++)
+  {
+
+    // Find small subdiagonal element
+    tst1 = std::max(tst1, std::fabs(d[l]) + std::fabs(e[l]));
+    int m = l;
+    while (m < N)
+    {
+      if (std::fabs(e[m]) <= eps * tst1)
+        break;
+      m++;
+    }
+
+    // If m == l, d[l] is an eigenvalue,
+    // otherwise, iterate.
+    if (m > l)
+    {
+      int iter = 0;
+      do
+      {
+        iter = iter + 1; // (Could check iteration count here.)
+
+        // Compute implicit shift
+        T g = d[l];
+        T p = (d[l + 1] - g) / (2.0 * e[l]);
+        T r = std::sqrt(p * p + 1.0);
+        if (p < 0)
+          r = -r;
+        d[l] = e[l] / (p + r);
+        d[l + 1] = e[l] * (p + r);
+        T dl1 = d[l + 1];
+        T h = g - d[l];
+        for (int i = l + 2; i < N; i++)
+          d[i] -= h;
+        f = f + h;
+
+        // Implicit QL transformation.
+        p = d[m];
+        T c = 1.0;
+        T c2 = c;
+        T c3 = c;
+        T el1 = e[l + 1];
```

Comment:
please check which variables can be marked `const`. This makes the code more explicit and may help the compiler perform further optimizations.